### PR TITLE
Improved opening and closing balloon panel

### PR DESCRIPTION
**Pull request by oskarwrobel**
I'm not proud of it, but this PR fixes few issues #20 #21 #26. Thy were strongly connected.

Related: https://github.com/ckeditor/ckeditor5-ui-default/issues/77

### CODE REVIEW

File: src/link.js [hunk 1]
```diff
@@ -178,40 +144,80 @@ export default class Link extends Feature {
...
+
+		// Handle `Ctrl+L` keystroke and show panel.
+		editor.keystrokes.set( 'CTRL+L', () => {
+			this._attachPanelToElement();
+			balloonPanel.urlInput.view.select();
 		} );
 
-		// Always focus editor on panel hide.
+		// Attach close by `Esc` press and click out of panel actions on panel show, on panel hide clean up listeners.
 		this.listenTo( balloonPanel.view.model, 'change:isVisible', ( evt, propertyName, value ) => {
```
Comment (oskarwrobel) [inline src/link.js]: We can consider moving this code to some helper and make it reusable.

File: src/link.js [hunk 2]
```diff
@@ -178,40 +144,80 @@ export default class Link extends Feature {
...
+					// Do nothing when the panel was clicked.
+					if ( balloonPanel.view.element.contains( domEvt.target ) ) {
+						return;
+					}
+
+					// When click was out of the panel then hide it.
+					balloonPanel.view.hide();
+
+					// When editor was clicked then restore editor focus.
+					if ( editor.ui.view.element.contains( domEvt.target ) ) {
```
Comment (oskarwrobel) [inline src/link.js]: It is bad and it is not tested. I know that features shouldn't depend on `EditorUI`, but we need to figure out how to restore focus on editable after clicking on some part of EditorUI e.g. toolbar button. Maybe EditorUI should do it itself? Listen to click on itself, ask `FocusManager` where is the focus, and restore if is needed? Or maybe there is an easier way which I don't see.

Comment (oskarwrobel) [inline src/link.js]: Lets forget about it :) I've removed this code.

**Comment by Reinmar** [comment]
From a functional perspective, this code works great. I could not find an issue in how it works.

The code, though, deserves a bigger refactoring already, so it's hard to do its review now. The refactoring is planned anyway, so let's postpone this until the next PR.

**Comment by Reinmar** [comment]
I'll run automated tests and if everything is OK, I'll merge the code.

**Comment by Reinmar** [comment]
There's one thing, other than https://github.com/ckeditor/ckeditor5-link/issues/40, which I'd like to improve. `_closePanelOnEsc` should not need to be protected and used in tests. Aren't there other ways to check that the "thing" happened?

**Comment by oskarwrobel** [comment]
I don't like it too. I'll try to find some other solution, but I've been trying and I haven't found.

**Comment by Reinmar** [comment]
I merged my t/20 fixes directly to master (https://github.com/ckeditor/ckeditor5-link/commit/8440ef2197c671ffa2671ac2e8f6e6b59f79096c) cause I failed to push them before merging this PR ;<

**Pull request merged**